casper/protocols/sharding/forkchoice.py

```diff
@@ -0,0 +1,88 @@
+"""The forkchoice module implements the estimator function a sharded blockchain"""
+
+
+def get_max_weight_indexes(scores):
+    """Returns the keys that map to the max value in a dict.
+    The max value must be greater than zero."""
+
+    max_score = max(scores.values())
+
+    assert max_score > 0, "max_score should be greater than zero"
+
+    max_weight_estimates = {e for e in scores if scores[e] == max_score}
+
+    return max_weight_estimates
+
+
+def get_scores(starting_block, latest_messages, shard_id):
+    """Returns a dict of block => weight"""
+    scores = dict()
+
+    for validator, current_block in latest_messages.items():
+        if not current_block.on_shard(shard_id):
+            continue
+
+        while current_block and current_block != starting_block:
+            scores[current_block] = scores.get(current_block, 0) + validator.weight
+            current_block = current_block.prev_block(shard_id)
```

Comment:
Can previous blocks occur in multiple shards? Or is this a unique selection, i.e. each block's previous history occurs in exactly 1 shard

Reply:
A merge-block has prev_block's of multiple shards (2 in this current implementation). This is the core mechanism that provides atomicity of shard transactions and builds up the fork choice rule in binary sharding.

Reply:
@djrtwo that makes sense, thanks!